File: Java/src/com/daily/algothrim/lookup/BinarySearch.java
package com.daily.algothrim.lookup;

/**
 * 二分查找
 */
public class BinarySearch {

    public static void main(String[] args) {
        int[] a = new int[]{1, 2, 4, 5, 6, 9, 29, 93, 95, 213, 741};
        System.out.println(new BinarySearch().search(a, a.length, 93));
    }

    /**
     * O(log n)
     */
    private int search(int[] a, int n, int value) {
        int start = 0;
        int end = n - 1;

        while (end >= start) {
            // 注意运算优先级
            int mid = start + ((end - start) >> 1);
            if (value == a[mid]) {
                return mid;
            } else if (value < a[mid]) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }

        return -1;
    }

}

File: kotlin/src/com/daily/algothrim/lookup/BinarySearch.kt
package com.daily.algothrim.lookup

/**
 * 二分查找
 */
class BinarySearch {
    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val a = intArrayOf(1, 2, 4, 5, 6, 9, 29, 93, 95, 213, 741)
            println(BinarySearch().search(a, a.size, 93))
        }
    }

    /**
     * O(log n)
     */
    fun search(a: IntArray, n: Int, value: Int): Int {
        var low = 0
        var high = n - 1

        while (low <= high) {
            val mid = low + (high - low).shr(1)
            when {
                a[mid] < value -> {
                    low = mid + 1
                }
                a[mid] > value -> {
                    high = mid - 1
                }
                else -> {
                    return mid
                }
            }
        }
        return -1
    }
}